libminifi/src/c2/C2Agent.cpp

```diff
@@ -873,6 +664,161 @@ void C2Agent::update_agent() {
+  if (protocol_.load() == nullptr) {
+    return {};
+  }
+
+  std::string resolved_url = uri;
+  if (!utils::StringUtils::startsWith(uri, "http")) {
+    std::stringstream adjusted_url;
+    std::string base;
+    if (configuration_->get(minifi::Configure::nifi_c2_flow_base_url, base)) {
+      adjusted_url << base;
+      if (!utils::StringUtils::endsWith(base, "/")) {
+        adjusted_url << "/";
+      }
+      adjusted_url << uri;
+      resolved_url = adjusted_url.str();
+    } else if (configuration_->get("c2.rest.url", base)) {
+      std::string host, protocol;
+      int port = -1;
+      utils::parse_url(&base, &host, &port, &protocol);
+      adjusted_url << protocol << host;
+      if (port > 0) {
+        adjusted_url << ":" << port;
+      }
+      adjusted_url << "/c2/api/" << uri;
+      resolved_url = adjusted_url.str();
+    }
+  }
+
+  C2Payload payload(Operation::TRANSFER, false, true);
+  C2Payload &&response = protocol_.load()->consumePayload(resolved_url, payload, RECEIVE, false);
+
+  auto raw_data = response.getRawData();
+  return std::string(raw_data.data(), raw_data.size());
+}
+
+bool C2Agent::handleConfigurationUpdate(const C2ContentResponse &resp) {
+  auto url = resp.operation_arguments.find("location");
+
+  std::string file_uri;
+  std::string configuration_str;
+
+  if (url == resp.operation_arguments.end()) {
+    logger_->log_debug("Did not have location within %s", resp.ident);
+    auto update_text = resp.operation_arguments.find("configuration_data");
+    if (update_text == resp.operation_arguments.end()) {
+      logger_->log_debug("Neither the config file location nor the data is provided");
+      C2Payload response(Operation::ACKNOWLEDGE, state::UpdateState::SET_ERROR, resp.ident, false, true);
+      response.setRawData("Error while applying flow. Neither the config file location nor the data is provided.");
+      enqueue_c2_response(std::move(response));
+      return false;
+    }
+    configuration_str = update_text->second.to_string();
+  }
+
+  file_uri = url->second.to_string();
+  utils::optional<std::string> optional_configuration_str = fetchFlow(file_uri);
+  if (!optional_configuration_str) {
+    logger_->log_debug("Couldn't load new flow configuration from: \"%s\"", file_uri);
+    C2Payload response(Operation::ACKNOWLEDGE, state::UpdateState::SET_ERROR, resp.ident, false, true);
+    response.setRawData("Error while applying flow. Couldn't load flow configuration.");
+    enqueue_c2_response(std::move(response));
+    return false;
+  }
+  configuration_str = optional_configuration_str.value();
+
+  bool should_persist = [&] {
+    auto persist = resp.operation_arguments.find("persist");
+    if (persist == resp.operation_arguments.end()) {
+      return false;
+    }
+    return utils::StringUtils::equalsIgnoreCase(persist->second.to_string(), "true");
+  }();
+
+  int16_t err = {update_sink_->applyUpdate(file_uri, configuration_str, should_persist)};
```

Comment:
why is there a `{ ... }` around this?

Reply:
to prevent accidental narrowing conversion in case we change the return type of `applyUpdate`